Test Failure: Failover1ServerCoordinatedPollingTest.testMultipleInstancesCompeteToRunManyLateTasksPC

Test Failure: com.ibm.ws.concurrent.persistent.fat.failover1serv.Failover1ServerCoordinatedPollingTest.testMultipleInstancesCompeteToRunManyLateTasksPC

```
testMultipleInstancesCompeteToRunManyLateTasksPC:junit.framework.AssertionFailedError: 2021-05-08-20:35:07:088 The response did not contain "[SUCCESS]".  Full output is:"
ERROR: Caught exception attempting to call test method testCancelTask on servlet failover1serv.web.Failover1ServerTestServlet
javax.persistence.PersistenceException: Exception [EclipseLink-4002] (Eclipse Persistence Services - 2.6.8.WAS-v20210426-a9becf7): org.eclipse.persistence.exceptions.DatabaseException
Internal Exception: java.sql.SQLTransactionRollbackException: A lock could not be obtained within the time requested
Error Code: 30000
Call: SELECT LOADER, MBITS, INAME, NEXTEXEC, RESLT, STATES, VERSION FROM WLPTASK WHERE ((ID = ?) AND (OWNR = ?))
    bind => [2 parameters bound]
Query: ReportQuery(referenceClass=Task sql="SELECT LOADER, MBITS, INAME, NEXTEXEC, RESLT, STATES, VERSION FROM WLPTASK WHERE ((ID = ?) AND (OWNR = ?))")
    at org.eclipse.persistence.internal.jpa.QueryImpl.getDetailedException(QueryImpl.java:382)
    at org.eclipse.persistence.internal.jpa.QueryImpl.executeReadQuery(QueryImpl.java:260)
    at org.eclipse.persistence.internal.jpa.QueryImpl.getResultList(QueryImpl.java:473)
    at com.ibm.ws.concurrent.persistent.db.DatabaseTaskStore.findById(DatabaseTaskStore.java:490)
    at com.ibm.ws.concurrent.persistent.internal.PersistentExecutorImpl.getStatus(PersistentExecutorImpl.java:964)
    at failover1serv.web.Failover1ServerTestServlet.testCancelTask(Failover1ServerTestServlet.java:63)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at componenttest.app.FATServlet.doGet(FATServlet.java:67)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1258)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:746)
    at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:443)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1227)
    at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1011)
    at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:75)
    at com.ibm.ws.webcontainer40.servlet.CacheServletWrapper40.handleRequest(CacheServletWrapper40.java:85)
    at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:938)
    at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1159)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:428)
    at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:387)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:566)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:500)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:360)
    at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:327)
    at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:167)
    at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:75)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:504)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:574)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:958)
    at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1047)
    at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:238)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: Exception [EclipseLink-4002] (Eclipse Persistence Services - 2.6.8.WAS-v20210426-a9becf7): org.eclipse.persistence.exceptions.DatabaseException
Internal Exception: java.sql.SQLTransactionRollbackException: A lock could not be obtained within the time requested
Error Code: 30000
Call: SELECT LOADER, MBITS, INAME, NEXTEXEC, RESLT, STATES, VERSION FROM WLPTASK WHERE ((ID = ?) AND (OWNR = ?))
    bind => [2 parameters bound]
Query: ReportQuery(referenceClass=Task sql="SELECT LOADER, MBITS, INAME, NEXTEXEC, RESLT, STATES, VERSION FROM WLPTASK WHERE ((ID = ?) AND (OWNR = ?))")
    at org.eclipse.persistence.exceptions.DatabaseException.sqlException(DatabaseException.java:340)
    at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.basicExecuteCall(DatabaseAccessor.java:684)
    at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.executeCall(DatabaseAccessor.java:560)
    at org.eclipse.persistence.internal.sessions.AbstractSession.basicExecuteCall(AbstractSession.java:2070)
    at org.eclipse.persistence.sessions.server.ServerSession.executeCall(ServerSession.java:586)
    at org.eclipse.persistence.sessions.server.ClientSession.executeCall(ClientSession.java:262)
    at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:262)
    at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:248)
    at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeSelectCall(DatasourceCallQueryMechanism.java:319)
    at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.selectAllRows(DatasourceCallQueryMechanism.java:731)
    at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectAllRowsFromTable(ExpressionQueryMechanism.java:2748)
    at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectAllReportQueryRows(ExpressionQueryMechanism.java:2685)
    at org.eclipse.persistence.queries.ReportQuery.executeDatabaseQuery(ReportQuery.java:852)
    at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:908)
    at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1142)
    at org.eclipse.persistence.queries.ReadAllQuery.execute(ReadAllQuery.java:460)
    at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1230)
    at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:2900)
    at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1866)
    at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1848)
    at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1813)
    at org.eclipse.persistence.internal.jpa.QueryImpl.executeReadQuery(QueryImpl.java:258)
Caused by: java.sql.SQLTransactionRollbackException: A lock could not be obtained within the time requested
    at org.apache.derby.client.am.SQLExceptionFactory.getSQLException(Unknown Source)
    at org.apache.derby.client.am.SqlException.getSQLException(Unknown Source)
    at org.apache.derby.client.am.ClientResultSet.next(Unknown Source)
    at com.ibm.ws.rsadapter.jdbc.WSJdbcResultSet.next(WSJdbcResultSet.java:2253)
    at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.processResultSet(DatabaseAccessor.java:760)
    at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.basicExecuteCall(DatabaseAccessor.java:657)
Caused by: ERROR 40XL1: A lock could not be obtained within the time requested
    at org.apache.derby.client.am.ClientResultSet.completeSqlca(Unknown Source)
    at org.apache.derby.client.net.NetResultSetReply.parseFetchError(Unknown Source)
    at org.apache.derby.client.net.NetResultSetReply.parseCNTQRYreply(Unknown Source)
    at org.apache.derby.client.net.NetResultSetReply.readFetch(Unknown Source)
    at org.apache.derby.client.net.ResultSetReply.readFetch(Unknown Source)
    at org.apache.derby.client.net.NetResultSet.readFetch_(Unknown Source)
    at org.apache.derby.client.net.NetResultSet.flowFetch(Unknown Source)
    at org.apache.derby.client.net.NetCursor.getMoreData_(Unknown Source)
    at org.apache.derby.client.am.Cursor.stepNext(Unknown Source)
    at org.apache.derby.client.am.Cursor.next(Unknown Source)
    at org.apache.derby.client.am.ClientResultSet.nextX(Unknown Source)
". expected:<0> but was:<1>
    at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:537)
    at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:495)
    at componenttest.topology.utils.HttpUtils.findStringInUrl(HttpUtils.java:472)
    at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:444)
    at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:414)
    at componenttest.topology.utils.FATServletClient.runTest(FATServletClient.java:83)
    at com.ibm.ws.concurrent.persistent.fat.failover1serv.Failover1ServerCoordinatedPollingTest.testMultipleInstancesCompeteToRunManyLateTasksPC(Failover1ServerCoordinatedPollingTest.java:226)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:197)
    at com...
```

pollForUnclaimedTasks has a bug where it leaves a task that it cannot run reserved under inMemoryTaskIds if the claimIfNotLocked succeeds but the commit of that operation fails.  Due to this, the persistent executor skips the task, reporting:

```
[5/8/21, 20:31:25:622 PDT] 0000004b id=2580bf35 ibm.ws.concurrent.persistent.internal.PersistentExecutorImpl 3 Found task 5 already scheduled
[5/8/21, 20:31:25:622 PDT] 0000004b id=2580bf35 ibm.ws.concurrent.persistent.internal.PersistentExecutorImpl 3 Found task 6 already scheduled
[5/8/21, 20:31:25:622 PDT] 0000004b id=2580bf35 ibm.ws.concurrent.persistent.internal.PersistentExecutorImpl 3 Found task 2 already scheduled
[5/8/21, 20:31:25:622 PDT] 0000004b id=2580bf35 ibm.ws.concurrent.persistent.internal.PersistentExecutorImpl 3 Found task 4 already scheduled
```

Here is a stack of the commit failing,
```
org.apache.derby.client.am.XaException: XA_RBTIMEOUT : Error executing a XAResource.end(), server returned XA_RBTIMEOUT.
	at org.apache.derby.client.net.NetXAResource.throwXAException(Unknown Source)
	at org.apache.derby.client.net.NetXAResource.end(Unknown Source)
	at com.ibm.ws.rsadapter.impl.WSRdbXaResourceImpl.end(WSRdbXaResourceImpl.java:417)
	at com.ibm.ejs.j2c.XATransactionWrapper.end(XATransactionWrapper.java:380)
	at com.ibm.ws.Transaction.JTA.JTAResourceBase.end(JTAResourceBase.java:218)
	at com.ibm.tx.jta.impl.RegisteredResources.sendEnd(RegisteredResources.java:941)
	at com.ibm.tx.jta.impl.RegisteredResources.distributeEnd(RegisteredResources.java:920)
	at com.ibm.tx.jta.impl.TransactionImpl.internalRollback(TransactionImpl.java:1627)
	at com.ibm.tx.jta.impl.TransactionImpl.internalRollback(TransactionImpl.java:1597)
	at com.ibm.tx.jta.impl.TransactionImpl.coreStage2CommitProcessing(TransactionImpl.java:865)
	at com.ibm.tx.jta.impl.TransactionImpl.stage2CommitProcessing(TransactionImpl.java:891)
	at com.ibm.tx.jta.impl.TransactionImpl.processCommit(TransactionImpl.java:771)
	at com.ibm.tx.jta.impl.TransactionImpl.commit(TransactionImpl.java:711)
	at com.ibm.tx.jta.impl.TranManagerImpl.commit(TranManagerImpl.java:165)
	at com.ibm.tx.jta.impl.TranManagerSet.commit(TranManagerSet.java:113)
	at com.ibm.ws.transaction.services.TransactionManagerService.commit(TransactionManagerService.java:293)
	at com.ibm.ws.concurrent.persistent.internal.PersistentExecutorImpl$PollingTask.pollForUnclaimedTasks(PersistentExecutorImpl.java:2484)
	at com.ibm.ws.concurrent.persistent.internal.PersistentExecutorImpl$PollingTask.run(PersistentExecutorImpl.java:2526)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:238)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: ERROR XN019: Error executing a XAResource.end(), server returned XA_RBTIMEOUT.
	at org.apache.derby.client.net.NetXAResource.xaRetValErrorAccumSQL(Unknown Source)
	... 23 more
```

--- a/dev/com.ibm.ws.concurrent.persistent/src/com/ibm/ws/concurrent/persistent/internal/PersistentExecutorImpl.java
+++ b/dev/com.ibm.ws.concurrent.persistent/src/com/ibm/ws/concurrent/persistent/internal/PersistentExecutorImpl.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2014, 2020 IBM Corporation and others.
+ * Copyright (c) 2014, 2021 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -2477,14 +2477,15 @@ public class PersistentExecutorImpl implements ApplicationRecycleComponent, DDLG
                         now = System.currentTimeMillis();
                         long claimUntilTime = (now > nextExecTime ? now : nextExecTime) + config.missedTaskThreshold * 1000;
 
+                        boolean claimedPendingCommit = false;
                         tranMgr.begin();
                         try {
-                            claimed = taskStore.claimIfNotLocked(taskId, version, claimUntilTime);
+                            claimedPendingCommit = taskStore.claimIfNotLocked(taskId, version, claimUntilTime);
                         } finally {
                             tranMgr.commit();
                         }
 
-                        if (claimed) {
+                        if (claimed = claimedPendingCommit) {
                             short mbits = (Short) result[1];
                             int txTimeout = (Integer) result[3];
                             InvokerTask task = new InvokerTask(PersistentExecutorImpl.this, taskId, nextExecTime, mbits, txTimeout);

